[Bugfix] Enable chunked-prefill and prefix cache with flash-attn backend

sighingnow opened the pull request:
As the flash-attn becomes the default attention backend on NV GPUs (thanks to vllm-flash-attention fork), it is possible to enable chunked prefill with prefix caching at the same time in vLLM.

FIX #xxxx (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>
<!-- inside this <details> section, markdown rendering does not work, so we use raw html here. -->
<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>


</details>

sighingnow commented:
Hi @rkooo567, could you please take a look at this pull request?

Thanks!

sighingnow commented:
> I make a try and output consistency cannot be guaranteed under some circumstances.

May I know more about the failed case? How could I reproduce the failure?

rkooo567 commented:
I will take a look at it tomorrow.

comaniac commented on vllm/worker/model_runner.py:
```diff
@@ -277,7 +277,8 @@ def add_seq_group(self, seq_group_metadata: SequenceGroupMetadata):
+            if (self.chunked_prefill_enabled and prefix_cache_hit
+                    and self.attn_backend.get_name() != "flash-attn"):
                 raise RuntimeError(
                     "chunked prefill cannot be used with prefix caching now.")
```
```
Chunked prefill and prefix caching can only be used simultaneously with flash-attn backend
```

rkooo567 replied:
Update the error message that it is only suppored by flash attn backend (also please write how to enable flash attn backend. I.e., `VLLM_ATTENTION_BACKEND=FLASH_ATTN`)

comaniac commented on vllm/worker/model_runner.py:
```diff
@@ -302,8 +303,25 @@ def add_seq_group(self, seq_group_metadata: SequenceGroupMetadata):
+                #       cached parts of the sequence.
+                #   - prefix_cache_len >= seq_len:
+                #     - it means the current partial sequence is fully cache
+                #       hited, and no further computation is needed.
```
```suggestion
                #       hit, and no further computation is needed.
```

comaniac reviewed (commented):
Overall LGTM. btw can you add a test case?

rkooo567 reviewed (commented):
Looks pretty clean! Can you add unit tests? I think it'd be great to have 1 for worker/test_model_runner.py (just to verify metadata) and 1 e2e test

sighingnow committed:
Enable chunked-prefill and prefix cache with flash-attn backend

Signed-off-by: Tao He <linzhu.ht@alibaba-inc.com>

sighingnow commented:
Hi @rkooo567 @comaniac, I have addressed the review comments and added a unittest to check the model inputs, could you help to take another look on this PR?

Thanks!

comaniac commented on vllm/worker/model_runner.py:
```diff
@@ -365,20 +365,45 @@ def _compute_for_prefix_cache_hit(
         inter_data.prefix_cache_hit = prefix_cache_hit
-        if self.chunked_prefill_enabled and prefix_cache_hit:
+        if (self.chunked_prefill_enabled and prefix_cache_hit
+                and self.attn_backend.get_name() != "flash-attn"):
             raise RuntimeError(
-                "chunked prefill cannot be used with prefix caching now.")
+                "Chunked prefill and prefix caching can only be used "
+                "simultaneously with flash-attn backend, try switching "
+                "to flash-attn backend by setting the environment variable "
+                "\"VLLM_ATTENTION_BACKEND=FLASH_ATTN\"")
```
`prepare_input` is now backend agnostic. Can you put this checker to `vllm/attention/backends/flashinfer.py` and `vllm/attention/backends/utils.py`? For example in utils.py:

```python
    def _add_seq_group(
            self, inter_data: "ModelInputForGPUBuilder.InterDataForSeqGroup",
            chunked_prefill_enabled: bool):
        is_prompt = inter_data.is_prompt
        block_tables = inter_data.block_tables
        computed_block_nums = inter_data.computed_block_nums

        if chunked_prefill_enabled and inter_data.prefix_cache_hit:
            raise RuntimerError(...)
```

comaniac commented on vllm/worker/model_runner.py:
```diff
@@ -365,20 +365,45 @@ def _compute_for_prefix_cache_hit(
-            inter_data.input_positions[seq_idx] = inter_data.input_positions[
-                seq_idx][context_len:]
+            prefix_cache_len = len(computed_block_nums) * self.block_size
+            # When prefix caching meets chunked prefill, we would be in
+            # the following three conditions:
+            #
+            #   - prefix_cache_len <= context_len:
+            #     - do normal chunked prefill and nothing special
+            #   - context_len < prefix_cache_len < seq_len:
+            #     - advance the context_len to seq_len to perform non-
+            #       cached parts of the sequence.
+            #   - prefix_cache_len >= seq_len:
+            #     - it means the current partial sequence is fully cache
+            #       hit, and no further computation is needed.
```
nit
```suggestion
            # When prefix caching meets chunked prefill, we would be in
            # one of the following three cases:
            #   - prefix_cache_len <= context_len:
            #     Do normal chunked prefill
            #   - context_len < prefix_cache_len < seq_len:
            #     Advance the context_len to seq_len to prefill non-
            #     cached parts of the sequence.
            #   - prefix_cache_len >= seq_len:
            #     The current partial sequence is fully cache hit,
            #     and no further computation is needed.
```

Another way is embedding the comments to each branch:

```python
if context_len < prefix_cache_len < seq_len:
    # prefix_cache_len <= context_len: Do normal chunked prefill
    ....
elif seq_len <= prefix_cache_len:
    # context_len < prefix_cache_len < seq_len: ...
    ...
else:
    # prefix_cache_len >= seq_len:: ...
    pass
```

I'm fine with either way

comaniac commented on tests/worker/test_model_runner.py:
```diff
@@ -381,3 +383,99 @@ def test_hybrid_batches(batch_size, enforce_eager, distributed_init):
+
+    seq_lens: List[int] = []
+    seq_group_metadata_list: List[SequenceGroupMetadata] = []
+    block_tables = {0: list(range(128))}
```
Can you note the block size here for convenience?

sighingnow replied:
Fixed。

comaniac commented on tests/worker/test_model_runner.py:
```diff
@@ -381,3 +383,99 @@ def test_hybrid_batches(batch_size, enforce_eager, distributed_init):
+        model_input.attn_metadata,
+    )
+
+    assert len(input_tokens) == 220
+    assert len(input_positions) == 220
```
Can you note how 220 was calculated?

sighingnow replied:
Addressed.

comaniac reviewed (commented):
Overall LGTM! Just some miner comments for comments and refactoring.

sighingnow commented:
Hi @Juelianqvq could you take another try on this PR to see if previous problems you meet can be reproduced now.

@comaniac Could you please help review changes in this PR again? thanks!

sighingnow commented:
These two failed test case is not caused by this PR.

sighingnow commented:
> > Hi @Juelianqvq could you take another try on this PR to see if previous problems you meet can be reproduced now.
> 
> @sighingnow Let me test it with our inner dataset consists of high concurreny under differenct scenerios next Monday and give you feedback based on that ASAP. Have you tested with blockmanager v2 yet?

Make sure you are working on vllm version with PR https://github.com/vllm-project/vllm/pull/7018 included.

> Have you tested with blockmanager v2 yet?

Haven't tested yet, but there should be no difference.

sighingnow commented:
> > Hi @Juelianqvq could you take another try on this PR to see if previous problems you meet can be reproduced now.
> 
> @sighingnow Let me test it with our inner dataset consists of high concurreny under differenct scenerios next Monday and give you feedback based on that ASAP. Have you tested with blockmanager v2 yet?

Make sure you are working on vllm version with PR https://github.com/vllm-project/vllm/pull/7018 included.

> Have you tested with blockmanager v2 yet?

Haven't tested yet, but there should be no difference.

rkooo567 commented:
@Juelianqvq is it possible to provide a simple unit tests for us?

rkooo567 reviewed (commented):
The PR looks good to me, but the latest comment concerns me if there are still some other issues. 

@sighingnow is there any good way to create an e2e test?

sighingnow commented:
> but still can generate garbage sometimes

@Juelianqvq Could you share the command/script that you can reproduce the failure? even if your dataset cannot be available.

> is there any good way to create an e2e test?

@rkooo567 An e2e test cannot help the concerns here as we cannot reproduce. The only condition that may lead to incorrect result is hash collision (two non-equal sequences share the same pages). To reproduce such error, it is highly dataset dependent.

sighingnow committed:
FIxes the common_computed_block_nums updating.

Signed-off-by: Tao He <linzhu.ht@alibaba-inc.com>

sighingnow commented:
> @sighingnow Chunked prefill is connected with chunk size which can be inferred from max-num-batched-tokens, so prompt length greater or less than this value should be included. The reason you have missed to keep at least 1 token for prefill is that you do not select greater prompt length which is always fine in 1-concurrency e2e, but it will cause assertion error in `max-query-len > 0` under pressure test.
> 
> Moreover, v2-block-manager is another story.

See also: https://github.com/vllm-project/vllm/issues/7619

sighingnow commented:
> > The block manager v1 case has been fixed by [099ab34](https://github.com/vllm-project/vllm/commit/099ab34268d8e84a19334b89bc7caaf940684b05)
> 
> @sighingnow Thanks for the great catch, have been tested with custom dataset and all sounds good!

Thanks for the feedback!
